GoMeet/src/main/webapp/js/vote-meeting-time.js

```diff
@@ -0,0 +1,174 @@
+      // so don't alert them, but also don't crash the program or stop execution either.
+      // Hence not throwing an error (which stops execution), but logging it.
+      console.error(generateErrorMessage(time) + " - MeetingTimeId: " + meetingTimeIds[i]);
+    } else {
+      time.id = meetingTimeIds[i]; // Store the meetingId with the time to identify them later.
+      // Add time Id to the votedTimes if logged in user has voted for it.
+      if (time.voters !== null && time.voters.includes(currentUser)) {
+        votedTimes.add(time.id);
+      }
+      timeData.push(time);
+    }
+  }
+  generateVoteTimeForm(timeData, currentUser, votedTimes);
+}
+
+/**
+ * Generates a voting form containing the MeetingTime timeData 
+ * in the 'vote-meeting-times' <div>. 
+ * Current logged-in user will use this form to vote on the 
+ * associated meeting times.
+ * @param {Array[Object]} timeData An array of timeData objects
+ * retrieved from the '/meeting-time' servlet via GET request, in 
+ * the format:
+ * {meetingTimeId: id, datetime: datetimeStr, voteCount: count, voters: [voters]}
+ */
+function generateVoteTimeForm(timeData, currentUser, votedTimes) {
+  let votingRight = true;
+  if (votedTimes.size === MAX_VOTES) {
+    votingRight = false; // User may not vote again if they have voted maximum times.
+  }
+  // Disable all vote buttons for users that have voted MAX_VOTES already.
+  let disabledState = !votingRight; 
+
+  sortTimesbyVotes(timeData);
+
+  let voteTable = document.getElementById('meeting-times-table');
+  
+  for (let i = 0; i < timeData.length; i++) {
+    let time = timeData[i];
+    let timeRow = voteTable.insertRow(i + 1);
+
+    let dateTimeCell = timeRow.insertCell(0);
+    dateTimeCell.innerText = time.datetime;
+
+    let voteCountCell = timeRow.insertCell(1);
+    voteCountCell.innerText = time.voteCount;
+
+    let votersCell = timeRow.insertCell(2);
+    votersCell.innerText = time.voters;
+
+    // Clicking on this button will cast a vote for this meeting time.
+    let voteFormCell = timeRow.insertCell(3);
+    let voteButton = document.createElement('button');
+    voteButton.innerText = 'Vote Time'
+    voteButton.id = time.id;
+    voteButton.onclick = function() {voteTime(time.id, currentUser)};
+    if (votedTimes.has(time.id)) {
+      // Always disable buttons of times the user has voted for already.
+      voteButton.disabled = true;
+    } else {
+      voteButton.disabled = disabledState;
+    }
+    voteFormCell.appendChild(voteButton);
+  }
+}
+
+/**
+ * Sorts the timeData objects given by their 'voteCount'
+ * property in descending order
+ * @param {Array[Object]} timeData an array of timeData objects
+ * retrieved from the '/meeting-time' servlet via GET request, in 
+ * the format:
+ * {meetingTimeId: id, datetime: datetimeStr, voteCount: count, voters: [voters]}
+ */
+function sortTimesbyVotes(timeData) {
```

Comment:
Perhaps it would be good to explain why you decided to sort the times client-side. :)

Reply:
Yep, added some comments to do that in the function header. But here's the explanation for easy reference:

Sorting the timeData objects client-side, because it is likely that not every single client-side use case of the timeData requires the data to be sorted in vote order. Hence, sorting the timeData server-side will add unnecessary additional complexity to the server code (and consequently the fetch() calls to the servlet).